pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -137,10 +141,29 @@ func (c *NPLController) Run(stopCh <-chan struct{}) {
 	for i := 0; i < numWorkers; i++ {
 		go wait.Until(c.Worker, time.Second, stopCh)
 	}
-
+	c.SyncRules(stopCh)
```

Comment:
`SyncRules` sounds like a blocking call. Perhaps you can remove the call of `go wait.Until(f, SyncRuleInterval, stopCh)` in `SyncRules` to here. e.g.
```
go wait.Until(c.SyncRules, syncRuleInterval, stopCh)
```

Reply:
Hi @xliuxu. I have a slightly different idea. I want to suggest that SyncRuleInterval become a member of NPLController, and we can do something like:

```
type NPLController struct {
  SyncRuleInterval time.Duration  // Need to set once before Run(). After calling Run() any changes to SyncRuleInterval is ignored.
}

func NewNPLController(..., syncRuleInterval time.Duration) *NPLController {
  c := NPLController{
    ..
    SyncRuleInterval: syncRuleInterval,
    ..
  }
}

func (c *NPLController) SyncRules(stopCh <-chan struct{}) {
  f := ...
  wait.Until(f, SyncRuleInterval, stopCh)
}

func (c *NPLController) Run(stopCh <-chan struct{}) {
  ...
  go c.SyncRules(stopCh)
}
```

In test case we can call `go nplCtrl.SyncRules(stopCh)`. This does not only check logic for recovering rule, but also checks if this is done in some specific period. The coverage is a bit larger.